src/main/java/org/codehaus/plexus/components/io/resources/PlexusIoFileResource.java

```diff
@@ -59,23 +65,41 @@ protected PlexusIoFileResource( @Nonnull File file, @Nonnull String name,
         this( file, name, attrs, null, null );
     }
 
-    @SuppressWarnings( "ConstantConditions" )
     PlexusIoFileResource( @Nonnull final File file, @Nonnull String name, @Nonnull PlexusIoResourceAttributes attrs,
                           final ContentSupplier contentSupplier, final InputStreamTransformer streamTransformer )
         throws IOException
     {
-        super( name, file.lastModified(), file.length(), file.isFile(), file.isDirectory(), file.exists() );
-        this.file = file;
+        this( file, name, attrs,
+                new FileAttributes( file, true ),
+                contentSupplier, streamTransformer );
+    }
 
+    PlexusIoFileResource( @Nonnull final File file, @Nonnull String name,
+                          @Nonnull PlexusIoResourceAttributes attrs, @Nonnull FileAttributes fileAttributes,
+                          final ContentSupplier contentSupplier, final InputStreamTransformer streamTransformer )
+            throws IOException
+    {
+        super( name, fileAttributes.getLastModifiedTime().toMillis(), fileAttributes.getSize(),
+                fileAttributes.isRegularFile(), fileAttributes.isDirectory(),
+                fileAttributes.isRegularFile() || fileAttributes.isDirectory() || fileAttributes.isSymbolicLink() || fileAttributes.isOther() );
+        this.file = file;
+        this.attributes = requireNonNull( attrs, "attributes is null for file " + file.getName() );
+        this.fileAttributes = requireNonNull( fileAttributes, "fileAttributes is null for file " + file.getName() );
         this.contentSupplier = contentSupplier != null ? contentSupplier : getRootContentSupplier( file );
 
         boolean hasTransformer = streamTransformer != null && streamTransformer != identityTransformer;
         InputStreamTransformer transToUse = streamTransformer != null ? streamTransformer : identityTransformer;
 
         dfos = hasTransformer && file.isFile() ? asDeferredStream( this.contentSupplier, transToUse, this ) : null;
-        if ( attrs == null )
-            throw new IllegalArgumentException( "attrs is null for file " + file.getName() );
-        this.attributes = attrs;
+    }
+
+    private static <T> T requireNonNull( T t, String message )
```

Comment:
Drop this for [`Object.](https://docs.oracle.com/javase/8/docs/api/java/util/Objects.html#requireNonNull-T-java.lang.String-)